### ISSUE TITLE
sokol_app Gamepads

### ISSUE THREAD

**Issue by alex-rass-88** [floooh/sokol#200]
Will there be gamepad support? It would be nice to support gamepads on desktop platforms.

**Comment by septag** [comment]
There is 'input' plugin that supports gamepads as well as touch/keyboard and mouse.
I know the documentation is currently lacking and there is no example for input, but you can checkout `include/rizz/input.h` for the API.
example usage is something like this:

```c
rizz_input_device gamepad = the_input->create_device(RIZZ_INPUT_DEVICETYPE_PAD);
// map a button to your custom value
const int a_button = 1;
the_input->map_bool(gamepad, RIZZ_INPUT_PADBUTTON_A, a_button);

if (the_input->get_bool(a_button)) {
    // button 'X' is pressed
}
```

you can also use `the_input->show_debugger(NULL)` to see all enumerated devices and some debug info.

Cheers

**Comment by floooh** [comment]
Gamepad support is not planned for sokol_app.h, but there will most likely be a separate header at some point (I need this for my 8-bit emulators too). The API will probably be similar to the Oryol gamepad functions:

https://github.com/floooh/oryol/blob/043683dcb3181beb64ae1c85ea76e4a4eb71c124/code/Modules/Input/Input.h#L78-L105

**Comment by alex-rass-88** [comment]
I'll try to write myself. Then add to the collection of sokol libraries.

**Comment by floooh** [comment]
That's fine, but note that I haven't made up my mind whether it makes sense for me to accept PRs for entirely new headers, that may result in too much work for reviewing and discussing the API and code. So please don't be upset when you're doing all the work and I'm not accepting the PR :)

**Comment by TheSpydog** [comment]
I'm curious why you don't plan to support gamepads in the sokol_app header? Especially since it already supports mouse/keyboard/touch input, it seems like game controllers would belong there too. :)

**Comment by floooh** [comment]
@TheSpydog it basically depends on how closely the gamepad code is to the window system glue code (e.g. if a platforms needs a window handle to obtain gamepad status it would make sense to put it into sokol_app.h, otherwise not).

When I tinkered with gamepad support in Oryol, that was surprisingly complicated, because of "raw buttons" versus "mapped buttons". There isn't agreement between gamepads between button and axis indices (so "button 0" on one gamepad might be an entirely different button than "button 0" on another gamepad). So some sort of of gamepad-specific mapping is required which makes the API around gamepads non-trivial, unless only a handful standard gamepads should be supported (like PS4 and Xbox gamepads). So basically because a gamepad API would have to be more complex than just "GetButton(int index)" and "GetAxis(int index)" I think a separate header would make more sense.

**Comment by MikeHart66** [comment]
Hi flooh, any progress on this?

**Comment by floooh** [comment]
Nope, next feature on my list is pointer-lock / mouse-capture in sokol_app.h.

Maybe something like Gainput (http://gainput.johanneskuhlmann.de/) could help out in the meantime?

**Comment by MikeHart66** [comment]
Yes, I saw that but was concerned that there was not update since 2018. Will give it a try.

**Comment by tomasorti** [comment]
It is a shame... It would be great to play your pacmac clone with gamepad.
Sokol rocks!

**Comment by roig** [comment]
This is the only thing I miss from SDL2 when I changed to sokol. Gamepad support.
The rest was perfect, PERFECT. Sokol rules.

**Comment by void256** [comment]
What we all could use in the meantime is this lib: https://github.com/ThemsAllTook/libstem_gamepad (Gamepad C API wrapper for OS X, Windows, and Linux).

I just spend a couple of days with it and it works just great! It has a straightforward and IMHO very nice minimal Gamepad API. (I only tested it on macOS so far but I assume that Linux and Windows work equally good).

It does not do any mapping of buttons and axis so you only get the raw data (basically everything what [floooh](https://github.com/floooh) mentioned [above](https://github.com/floooh/sokol/issues/200#issuecomment-590892855)). So you need to do some kind of "I know this specific vendor and product id and axis 0 on this device is the left analog stick horizontal movement" but it's not too bad to add that on your own or add some kind of UI for people to add their own controllers.

From an API perspective it matches Sokol quite nicely. It would be nice to have it as a header only version tho. I guess it could be made into one but for now I'm fine of linking statically with it.

**Sokol indeed rocks very hard!** Lots of Kudos to master floooh for creating it! 🙇‍♂️

**Comment by void256** [comment]
Well, here I am, some four months later admitting that this is much harder than expected - at least when using a low level library like [libstem_gamepad](https://github.com/ThemsAllTook/libstem_gamepad) 😱 

The library itself is quite nice but the controllers are too different. I have three different ones here: a PS3 controller, a PS4 controller and a Nimbus Steel Series .. and there is a different issue with each one I tried :smile:

- The PS3 controller works fine - but it wants some some death zones for the analog axis which is not a problem.
- The Nimbus uses analog buttons where you get a value between -1 and 1 depending on how hard you press the buttons. Yes, that's something one can probably work around but probably not reliable because maybe on some other controller you'll get 0 to 1 or so, who knows.
- The PS4 controller returned some always increasing value for one of the axis which I couldn't figure out if it's a problem with libstem_gamepad intepreting some values wrong or something else (I tried two different PS4 controllers even and they both had that weird issue).

Well, to make a long story short, in the end I gave up on libstem_gamepad 😞 and I'm now using SDL2 Input (`SDL_Init(SDL_INIT_GAMECONTROLLER)`) together with Sokol.

In the end this worked so much better 😎 and because of the SDL2 controller database much more predictable and easy to use too.

🤓

**Comment by edubart** [comment]
I really miss a cross platform single header C library for gamepads. There is no such thing out there? Have anyone ever found one? Using SDL is not an option for me.
